hypothesis-python/src/hypothesis/strategies/_internal/types.py

```diff
@@ -407,6 +407,22 @@ def from_typing_type(thing):
         # the ghostwriter than it's worth, via undefined names in the repr.
         mapping.pop(collections.deque, None)
         mapping.pop(typing.Deque, None)
+    # namedtuples are, strictly speaking, generic. However, users treat them as
+    # dataclasses or structs, not as a generically typed collection, so we
+    # don't want to provide generic instantiations of these.
+    #
+    # Namedtuples aren't an actual type, so we'll check for internal attributes
+    # set by collections.namedtuple and hope nobody sets these on their own
+    # tuple subclass.
+    for t in sorted(mapping, key=type_sorting_key):
+        if (
+            isinstance(t, type)
+            and issubclass(t, tuple)
+            and hasattr(t, "_fields")
+            and hasattr(t, "_asdict")
+        ):
+            mapping.pop(t)
+
```

Comment:
I'm not sure where this should appear in the function order-wise. There's a portion lower down where `mapping` is iterated over in the same way - maybe this check belongs there?.

Reply:
On line 395 above it drops things which are exactly type tuple, maybe we could expand that to drop any subtype of tuple too?

Reply:
Seems reasonable, yeah.  I think we'd want to discard `os.Environ` first, and then _if there are non-tuple types_ discard all the tuple types.

Reply:
I've updated as you described: if there is at least one other type, drop all tuple (sub)classes.

I admit I'm not sure what input to `st.from_type` would cause `mapping` to be filled with *only* tuple subclasses. This might be relevant if the `cover` tests yell at me for not providing an input which fails the new conditional branch 🙂

Reply:
I think we need something along the lines of:
```python
import abc

class AbstractStruct(abc.ABC, tuple):
    @abstractmethod
    def foo(self): ...

class ConcreteA(AbstractStruct):
    def foo(self): pass
class ConcreteB(AbstractStruct):
    def foo(self): pass

def test_handles odd tuple_subclasses():
    s = st.from_type(AbstractStruct)
    assert_all_examples(s, lambda x: isinstance(x, AbstractStruct))
    find_any(s, lambda x: isinstance(x, ConcreteA))
    find_any(s, lambda x: isinstance(x, ConcreteB))
```

Reply:
I don't think this works, because `AbstractStruct` doesn't pass this conditional https://github.com/tybug/hypothesis/blob/f523db159e0f80163a3b3744240d73cb6b9445f2/hypothesis-python/src/hypothesis/strategies/_internal/core.py#L1364-L1369.

I managed to come up with a test involving a custom protocol whose only concrete instantiations are tuple subtypes, though - let me know what you think.